### PR TITLE
Generic MQTT improvements

### PR DESCRIPTION
**Pull request by relu91**
Given the latest problems that MQTT gave, I tried to revamp it a little bit the implementation in the hope that it would cause fewer problems and would be more maintainable. I took three steps:
- Upgrade the library to the latest MQTT implementation -> we get `async` methods -> cleaner implementation and cover corner cases where we forgot to use callbacks. 
- While I was there I updated the notation with the latest MQTT vocabulary
- I tried to tackle the issue raised in #955 by @ilovemilk . Now it should be fixed but give the current design of the Scripting API (and node-wot). the implementation is not that clean. I refer to the fact that we can't keep track of each listener by using functions as handlers but we use forms as identifiers. The end result is that if in your thing description you have two forms from different events that share a `mqv:filter` you can't subscribe to two events at the same time. I hope this is a corner case... if anybody else has suggestions on this let me know! 

Meanwhile I hope that this new changes also solves the problems in MacOS

### CODE REVIEW

**Comment by relu91** [comment]
> Meanwhile I hope that this new changes also solves the problems in MacOS

It seems it didn't let me try to play with it little more...

**Comment by relu91** [comment]
In b79e59b I tried to force qos 2 for every event. I suspect that the test is falling due to the fact that by default the event is published with qos 0 and therefore we lost some events. If it works I then try to improve the solution and make it more generic.

**Comment by relu91** [comment]
Ok now is failing due to file client (which fails also locally). @JKRhb can you fix in a parallel PR?

**Comment by relu91** [comment]
Ah maybe I **did see it** correctly:
image

File: packages/binding-file/test/test2.txt
Comment (relu91) [reply packages/binding-file/test/test2.txt:1]: Well spotted! I'll investigate.

**Comment by relu91** [comment]
I rebased with the master to start on the changes that @JKRhb made. Let's wait for the final greenlight from the CI/CD and @egekorkan.

**Comment by relu91** [comment]
Ok CI/CD is still angry. The funny thing is that I have a Windows setup with Node 18.x and even if I run the tests sequentially like the following:
```bash
for i in {1..100}; do npm run test -w packages/binding-mqtt; done
```
I get 100% success rate. I think we need more information if we want to know more about why is falling. Could it be that it finds some occupied port and just waits there? Currently, we are firing 5 events and the code is waiting for 3... with qos 2 it is possible that we lost so many packages (they shouldn't be lost at all!). 

Any idea how to debug better? Shall we activate `DEBUG=node-wot:binding-mqtt:*` on the CI/CD?

**Comment by JKRhb** [comment]
> Ok CI/CD is still angry. The funny thing is that I have a Windows setup with Node 18.x and even if I run the tests sequentially like the following:

Hmm, that's very strange :/

> Any idea how to debug better? Shall we activate `DEBUG=node-wot:binding-mqtt:*` on the CI/CD?

That sounds good! Maybe we can somehow configure the CI to re-run failed workflow runs with logging enabled in general? Could be useful to determine the source of problem more quickly.

**Review by egekorkan — APPROVED** [review]
Thanks a lot @relu91. The changes are many but all on the right direction!

**Comment by relu91** [comment]
> That sounds good! Maybe we can somehow configure the CI to re-run failed workflow runs with logging enabled in general? Could be useful to determine the source of problems more quickly. Edit: Just found an Action that could be used for that: https://github.com/marketplace/actions/retry-step#run-different-command-after-first-failure

It could be handy yes, however, I'd probably need to enable debug for the whole node-wot package (i.e. `DEBUG=node-wot`). What's your feeling about it? does it work?

**Comment by JKRhb** [comment]
> > That sounds good! Maybe we can somehow configure the CI to re-run failed workflow runs with logging enabled in general? Could be useful to determine the source of problems more quickly. Edit: Just found an Action that could be used for that: https://github.com/marketplace/actions/retry-step#run-different-command-after-first-failure
> 
> It could be handy yes, however, I'd probably need to enable debug for the whole node-wot package (i.e. `DEBUG=node-wot`). What's your feeling about it? does it work?

I've just opened https://github.com/eclipse-thingweb/node-wot/pull/1189 as an experimental PR to test it out :)

**Comment by danielpeintner** [comment]
> Thanks a lot @relu91. The changes are many but all on the right direction!

@relu91 Once you feel comfortable I think you can merge!

**Comment by relu91** [comment]
Ok, I review the current status and I think we can merge it. There is still the problem of how properly generate forms in the mqtt broker following user preference (i.e. how the user can configure `qos` when exposing a Thing). We can improve it later and as it has also some implications with https://github.com/eclipse-thingweb/node-wot/issues/1029.  Also thanks to #1189 we can debug better what's happening.

**Pull request merged**
